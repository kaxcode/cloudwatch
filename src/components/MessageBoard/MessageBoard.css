.MessageBoard__Input {
  grid-area: message;
  padding: 30px 30px 0 30px;
}

.MessageBoard__Output {
  position: absolute;
  z-index: 1000;
  margin: 0 auto;
  text-align: center;
  font-family: "Baloo Bhai", cursive;
  background-color: grey;
}
