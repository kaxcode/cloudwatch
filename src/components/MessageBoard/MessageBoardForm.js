import React from 'react';
import { func } from 'prop-types';
import './MessageBoard.css';

const MessageOutput = props => {
  if (window.name === 'presenter') {
    return null;
  }

  return (
    <form id="message-submit-form" onSubmit={props.onSubmit}>
      <input id="message-input" onChange={props.onChange} />
      <button id="message-submit-button">Post Message</button>
      <button onClick={props.onPresent}>Open Presentation Window</button>
    </form>
  );
};

MessageOutput.propTypes = {
  onChange: func.isRequired,
  onSubmit: func.isRequired,
  onPresent: func.isRequired
};

export default MessageOutput;
