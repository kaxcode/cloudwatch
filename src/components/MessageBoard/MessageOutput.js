import React from 'react';
import { func, bool } from 'prop-types';
import './MessageBoard.css';
import PropTypes from 'prop-types';

const MessageOutput = props => {
  if (!props.showMessage) {
    return null;
  }
  if (localStorage.message === '') {
    return null;
  }
  return (
    <div className="MessageBoard__Output">
      <h1>{props.message}</h1>
      <button onClick={props.handleDismiss}>X</button>
    </div>
  );
};

MessageOutput.propTypes = {
  showMessage: bool.isRequired,
  handleDismiss: func.isRequired,
  message: PropTypes.oneOfType([PropTypes.string, PropTypes.object])
};

export default MessageOutput;
