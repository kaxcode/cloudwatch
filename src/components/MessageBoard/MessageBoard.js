import React from 'react';
import MessageOutput from './MessageOutput';
import './MessageBoard.css';
import { object, number } from 'prop-types';
import MessageBoardForm from './MessageBoardForm';

class MessageBoard extends React.Component {
  state = {
    showMessage: false,
    message: ''
  };

  componentDidMount = () => {
    // check if window is parent or presenter
    if (window.name === 'presenter') {
      this.setState({ message: localStorage.message });
    }
  };

  handleDismiss = () => {
    this.setState({ showMessage: false });
    localStorage.setItem('message', '');
  };

  handleChange = e => {
    this.setState({ message: e.target.value });
  };

  handleSubmit = e => {
    e.preventDefault();
    this.setState({ showMessage: true });
    localStorage.setItem('message', this.state.message);
  };

  handlePresent = () => {
    localStorage.setItem('message', this.state.message);
      this.props.location.pathname === '/timer'
      ? localStorage.setItem('timeRemaining', this.props.timeRemaining)
      : localStorage.setItem('counter', this.props.counter);
    window.open('present', 'presenter');
  };

  render() {
    return (
      <div className="MessageBoard__Input">
        <MessageOutput
          message={this.state.message}
          showMessage={this.state.showMessage}
          handleDismiss={this.handleDismiss}
        />
        <MessageBoardForm
          onSubmit={this.handleSubmit}
          onChange={this.handleChange}
          onPresent={this.handlePresent}
        />
        <MessageOutput
          message={localStorage.message || ''}
          showMessage
          handleDismiss={this.handleDismiss}
        />
      </div>
    );
  }
}

MessageBoard.propTypes = {
  location: object,
  timeRemaining: number,
  counter: number
};

export default MessageBoard;
