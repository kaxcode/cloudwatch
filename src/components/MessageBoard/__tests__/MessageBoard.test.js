import React from 'react';
import Enzyme from 'enzyme';
import MessageBoard from '../MessageBoard.js';
import { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({ adapter: new Adapter() });

class LocalStorageMock {
  constructor() {
    this.message = {};
  }

  clear() {
    this.message = {};
  }

  getItem(key) {
    return this[key] || null;
  }

  setItem(key, value) {
    this[key] = value.toString();
  }

  removeItem(key) {
    delete this[key];
  }
}

global.localStorage = new LocalStorageMock();

describe('MessageBoard', () => {
  describe('#handleDismiss', () => {
    it('does not render a message when #handleDismiss is called', () => {
      //Arrange
      const subject = shallow(<MessageBoard />);
      //Act
      subject.instance().handleDismiss();
      //Assert
      expect(subject.state().message).toEqual('');
    });
  });

  describe('#componentDidMount', () => {
    it('checks if window is in presenter mode and if so sets the message state to that in localStorage', () => {
      // Arrange
      const testMessage = 'testMessage';
      const subject = shallow(
        <MessageBoard
          location={{ pathname: '/timer' }}
          timeRemaining={5}
          counter={5}
        />
      );
      global.name = 'presenter';
      // Act
      localStorage.setItem('message', testMessage);
      subject.instance().componentDidMount();
      // Assert
      expect(subject.state().message).toEqual(testMessage);
    });
  });

  describe('#handleSubmit', () => {
    it('renders a message when showMessage is TRUE', () => {
      //Arrange
      const subject = shallow(
        <MessageBoard
          location={{ pathname: '/timer' }}
          timeRemaining={5}
          counter={5}
        />
      );
      //Act
      subject.setState({ showMessage: true });
      //Assert
      expect(subject.state().showMessage).toBe(true);
    });
    it('calls #preventDefault when called', () => {
      const e = { preventDefault: jest.fn() };
      const subject = shallow(
        <MessageBoard
          location={{ pathname: '/timer' }}
          timeRemaining={5}
          counter={5}
        />
      );
      subject.instance().handleSubmit(e, 'test');
      expect(e.preventDefault).toHaveBeenCalled();
    });
  });

  describe('#handlePresent', () => {
    it('sets the message in localStorage', () => {
      //Arrange
      const subject = shallow(
        <MessageBoard
          location={{ pathname: '/timer' }}
          timeRemaining={5}
          counter={5}
        />
      );
      // Act
      localStorage.setItem('message', 'test message');
      subject.setState({ message: 'test message' });
      // Act
      subject.update();
      // Assert
      expect(localStorage.getItem('message')).toEqual('test message');
    });
  });

  it('opens secondary window with location and assigns the name presenter to it', () => {
    localStorage.setItem('message', 'test message');
    const subject = shallow(
      <MessageBoard
        location={{ pathname: '/timer' }}
        timeRemaining={5}
        counter={5}
      />
    );
    subject.setState({ message: 'test message' });
    window.open = jest.fn();
    subject.instance().handlePresent();
    expect(window.open).toBeCalled();
  });
});
