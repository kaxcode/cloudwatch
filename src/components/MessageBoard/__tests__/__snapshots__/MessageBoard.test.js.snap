// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageBoard checks whether the user is coming from /timer or /stopwatch 1`] = `
<div
  className="MessageBoard__Input"
>
  <MessageOutput
    handleDismiss={[Function]}
    message="test message"
    showMessage={false}
  />
  <MessageOutput
    onChange={[Function]}
    onPresent={[Function]}
    onSubmit={[Function]}
  />
  <MessageOutput
    handleDismiss={[Function]}
    message="test message"
    showMessage={true}
  />
</div>
`;

exports[`MessageBoard checks whether the user is coming from /timer or /stopwatch 2`] = `
<div
  className="MessageBoard__Input"
>
  <MessageOutput
    handleDismiss={[Function]}
    message="test message"
    showMessage={false}
  />
  <MessageOutput
    onChange={[Function]}
    onPresent={[Function]}
    onSubmit={[Function]}
  />
  <MessageOutput
    handleDismiss={[Function]}
    message="test message"
    showMessage={true}
  />
</div>
`;
